vllm/spec_decode/multi_step_worker.py

```diff
@@ -66,22 +66,10 @@ def sampler_output(
         copied_execute_model_req = execute_model_req.clone(
             copied_seq_group_metadata_list)
 
-        # Assert enough KV space for sample_len tokens per sequence.
-        self._assert_enough_kv_space(execute_model_req.seq_group_metadata_list,
-                                     sample_len)
-
         # Run model sample_len times.
-        model_outputs: List[SamplerOutput] = []
-        for _ in range(sample_len):
-            model_output = super().execute_model(
-                execute_model_req=copied_execute_model_req)
-            assert (len(model_output) == 1
-                    ), "composing multistep workers not supported"
-            model_output = model_output[0]
-
-            self._append_new_tokens(model_output,
-                                    copied_seq_group_metadata_list)
-            model_outputs.append(model_output)
+        copied_execute_model_req.num_steps = sample_len
```

Comment:
As per the offline discussions on slack this is my understanding of how we will plugin the bonus token logic here. Please confirm if my understanding is correct

```
expanded_request, original_seq_indices = self.expand_batch(
    copied_execute_model_req, seq_ids_with_bonus_tokens)
model_outputs: List[SamplerOutput] = self.execute_model(
            execute_model_req=expanded_request)
filtered_model_output = filter_output(model_outputs, original_seq_indices)
.....
.....

def expand_batch(
    self, copied_execute_model_req, seq_ids_with_bonus_tokens):
"""
Takes the original request and expands it for sequences which have bonus tokens. Returns
a new request with the expanded batch and also the indices of the original sequences.
"""
....
....


def filter_output(
    self, model_outputs, original_seq_indices):
"""
Iterate over model_outputs and filters the output of each step. 
The filtering retains the output corresponding to indices specified in original_seq_indices.
Returns the filtered output.

"""
.....
.....

```

Reply:
Yes this is also my understanding. One miner point to clarify is we will put the num_steps:

```python
model_outputs: List[SamplerOutput] = self.execute_model(
    execute_model_req=expanded_request, num_steps=num_steps
)
```